### ISSUE TITLE
Sema: `if case _ as` casting pattern doesn't work despite being implemented already

### ISSUE THREAD

**Issue by simanerush** [apple/swift#60808]
**Steps To Reproduce**
For example, this code snippet 
```swift
protocol T {}
struct A: T {}

func test_switch<T>(arr: T) {
  if case _ as A = arr {}
}
```
should work fine. However, it gives the following error: 
```
error: cannot convert value of type 'T' to specified type 'A'
  if case _ as A = arr {}
                   ^~~
                       as! A
```
and emits non-relevant diagnostics. 

**Expected behavior**
Even though this works by binding the result of the cast to a constant (`if case let t as A = arr {}`), the case specified above should also work.

**Environment (please fill out the following information)**
 - OS: macOS 12.5.1 
 - Swift version: 5.8-dev

**Comment by CodaFi** [comment]
The workaround pattern here is `if case is A = arr`

**Comment by simanerush** [comment]
@CodaFi So a version of this with the `as` pattern is not intended to be working in the first place?

**Comment by AnthonyLatsis** [comment]
Robert is simply pointing out that `case is A = foo`, which is semantically equivalent, can be used to work around the issue with `case _ as A = foo`; both are expected to work.

**Comment by hamishknight** [comment]
This was fixed by #64280, and should therefore be fixed in Swift 5.9 with Xcode 15 beta. Adding a test case in #68035.
